File: modules/agent/src/services/contact_service.py
"""Service layer for contact business logic."""

import logging
from typing import Optional, List, Dict, Any

from fastapi import HTTPException

from repositories.contact_repository import (
    search_contacts_and_individuals,
    delete_contact as delete_contact_repo,
    find_all_contacts_paginated,
    find_contact_by_id,
    create_contact_with_accounts,
    update_contact_with_accounts,
    ContactCreate,
    ContactUpdate,
)

# Re-export Pydantic models for controllers
__all__ = ["ContactCreate", "ContactUpdate"]

logger = logging.getLogger(__name__)


async def get_contacts_paginated(
    page: int,
    page_size: int,
    order_by: str,
    order: str,
    search: Optional[str] = None,
):
    """Get contacts with pagination."""
    return await find_all_contacts_paginated(
        page=page,
        page_size=page_size,
        search=search,
        order_by=order_by,
        order=order,
    )


async def search_contacts(query: str, tenant_id: Optional[int]):
    """Search contacts and individuals by name or email."""
    return await search_contacts_and_individuals(query, tenant_id=tenant_id)


async def get_contact(contact_id: int):
    """Get contact by ID."""
    contact = await find_contact_by_id(contact_id)
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


async def create_contact(
    contact_data: Dict[str, Any],
    user_id: Optional[int],
    account_ids: Optional[List[int]],
):
    """Create a contact with optional account links."""
    data = {
        "first_name": contact_data.get("first_name"),
        "last_name": contact_data.get("last_name"),
        "email": contact_data.get("email"),
        "phone": contact_data.get("phone"),
        "title": contact_data.get("title"),
        "department": contact_data.get("department"),
        "role": contact_data.get("role"),
        "notes": contact_data.get("notes"),
        "is_primary": contact_data.get("is_primary", False),
        "created_by": user_id,
        "updated_by": user_id,
    }
    return await create_contact_with_accounts(data, account_ids)


async def update_contact(
    contact_id: int,
    contact_data: Dict[str, Any],
    user_id: Optional[int],
    account_ids: Optional[List[int]],
):
    """Update a contact with account links."""
    data = dict(contact_data)
    data["updated_by"] = user_id

    contact = await update_contact_with_accounts(contact_id, data, account_ids)
    if not contact:
        raise HTTPException(status_code=404, detail="Contact not found")
    return contact


async def delete_contact(contact_id: int):
    """Delete a contact."""
    success = await delete_contact_repo(contact_id)
    if not success:
        raise HTTPException(status_code=404, detail="Contact not found")
    return True

File: modules/agent-go/internal/services/contact_service.go
package services

import (
	"errors"

	"github.com/pina-colada-co/agent-go/internal/repositories"
	"github.com/pina-colada-co/agent-go/internal/schemas"
	"github.com/pina-colada-co/agent-go/internal/serializers"
)

var ErrContactNotFound = errors.New("contact not found")
var ErrContactNameRequired = errors.New("first_name and last_name are required")

// ContactService handles contact business logic
type ContactService struct {
	contactRepo *repositories.ContactRepository
}

// NewContactService creates a new contact service
func NewContactService(contactRepo *repositories.ContactRepository) *ContactService {
	return &ContactService{contactRepo: contactRepo}
}

// GetContacts returns paginated contacts
func (s *ContactService) GetContacts(page, pageSize int, orderBy, order, search string, tenantID *int64) (*serializers.PagedResponse, error) {
	params := repositories.NewPaginationParams(page, pageSize, orderBy, order)
	params.Search = search

	result, err := s.contactRepo.FindAll(params, tenantID)
	if err != nil {
		return nil, err
	}

	items := make([]serializers.ContactResponse, len(result.Items))
	for i := range result.Items {
		items[i] = serializers.ContactToResponse(&result.Items[i])
	}

	resp := serializers.NewPagedResponse(items, result.TotalCount, result.Page, result.PageSize, result.TotalPages)
	return &resp, nil
}

// GetContact returns a contact by ID
func (s *ContactService) GetContact(id int64) (*serializers.ContactResponse, error) {
	contact, err := s.contactRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if contact == nil {
		return nil, ErrContactNotFound
	}

	resp := serializers.ContactToResponse(contact)
	return &resp, nil
}

// CreateContact creates a new contact
func (s *ContactService) CreateContact(input schemas.ContactCreate, userID int64) (*serializers.ContactResponse, error) {
	if input.FirstName == "" || input.LastName == "" {
		return nil, ErrContactNameRequired
	}

	repoInput := repositories.ContactCreateInput{
		UserID:     userID,
		FirstName:  input.FirstName,
		LastName:   input.LastName,
		Email:      input.Email,
		Phone:      input.Phone,
		Title:      input.Title,
		Department: input.Department,
		Role:       input.Role,
		Notes:      input.Notes,
		IsPrimary:  input.IsPrimary,
		AccountIDs: input.AccountIDs,
	}

	contactID, err := s.contactRepo.Create(repoInput)
	if err != nil {
		return nil, err
	}

	return s.GetContact(contactID)
}

// UpdateContact updates an existing contact
func (s *ContactService) UpdateContact(id int64, input schemas.ContactUpdate, userID int64) (*serializers.ContactResponse, error) {
	contact, err := s.contactRepo.FindByID(id)
	if err != nil {
		return nil, err
	}
	if contact == nil {
		return nil, ErrContactNotFound
	}

	if err := s.applyContactUpdates(id, input, userID); err != nil {
		return nil, err
	}

	if err := s.updateAccountLinks(id, input, contact.IsPrimary); err != nil {
		return nil, err
	}

	return s.GetContact(id)
}

func (s *ContactService) applyContactUpdates(id int64, input schemas.ContactUpdate, userID int64) error {
	updates := buildContactUpdates(input, userID)
	if len(updates) <= 1 {
		return nil
	}
	return s.contactRepo.Update(id, updates)
}

func (s *ContactService) updateAccountLinks(id int64, input schemas.ContactUpdate, currentIsPrimary bool) error {
	if input.AccountIDs == nil {
		return nil
	}
	isPrimary := resolvePrimaryFlag(input.IsPrimary, currentIsPrimary)
	return s.contactRepo.LinkToAccounts(id, input.AccountIDs, isPrimary)
}

func resolvePrimaryFlag(inputFlag *bool, current bool) bool {
	if inputFlag != nil {
		return *inputFlag
	}
	return current
}

// DeleteContact deletes a contact
func (s *ContactService) DeleteContact(id int64) error {
	contact, err := s.contactRepo.FindByID(id)
	if err != nil {
		return err
	}
	if contact == nil {
		return ErrContactNotFound
	}

	return s.contactRepo.Delete(id)
}

// SearchContacts searches contacts by name or email
func (s *ContactService) SearchContacts(query string, limit int) ([]serializers.ContactResponse, error) {
	if limit <= 0 {
		limit = 20
	}

	contacts, err := s.contactRepo.Search(query, limit)
	if err != nil {
		return nil, err
	}

	results := make([]serializers.ContactResponse, len(contacts))
	for i := range contacts {
		results[i] = serializers.ContactToResponse(&contacts[i])
	}

	return results, nil
}

func buildContactUpdates(input schemas.ContactUpdate, userID int64) map[string]interface{} {
	updates := map[string]interface{}{
		"updated_by": userID,
	}

	if input.FirstName != nil {
		updates["first_name"] = *input.FirstName
	}
	if input.LastName != nil {
		updates["last_name"] = *input.LastName
	}
	if input.Email != nil {
		updates["email"] = *input.Email
	}
	if input.Phone != nil {
		updates["phone"] = *input.Phone
	}
	if input.Title != nil {
		updates["title"] = *input.Title
	}
	if input.Department != nil {
		updates["department"] = *input.Department
	}
	if input.Role != nil {
		updates["role"] = *input.Role
	}
	if input.Notes != nil {
		updates["notes"] = *input.Notes
	}
	if input.IsPrimary != nil {
		updates["is_primary"] = *input.IsPrimary
	}

	return updates
}
